app/lib/search/search_service.dart

```diff
@@ -418,6 +434,56 @@ class SearchQuery {
   }
 }
 
+/// Filter conditions on tags.
+class TagsPredicate {
+  /// tag -> required (true) | forbidden (false) map
+  final _values = <String, bool>{};
+
+  TagsPredicate([Map<String, bool> values]) {
+    if (values != null) _values.addAll(values);
+  }
+
+  /// Parses [values] passed via Uri.queryParameters
+  TagsPredicate.parseQueryValues(List<String> values) {
```

Comment:
Can we annotate this `@visiblelForTesting` and consider it for a few other methods too?

Reply:
This is used in `SearchQuery.fromServiceUrl`.